packages/next/src/build/webpack/loaders/next-barrel-loader.ts

```diff
@@ -86,35 +86,177 @@
+
+    const matches = output.match(
+      /^([^]*)export (const|var) __next_private_export_map__ = ('[^']+'|"[^"]+")/
+    )
+    if (!matches) {
+      return null
+    }
+
+    const prefix = matches[1]
+    let exportList = JSON.parse(matches[3].slice(1, -1)) as [
+      string,
+      string,
+      string
+    ][]
+    const wildcardExports = [
+      ...output.matchAll(/export \* from "([^"]+)"/g),
+    ].map((match) => match[1])
+
+    // In the wildcard case, if the value is exported from another file, we
+    // redirect to that file (decl[0]). Otherwise, export from the current
+    // file itself.
+    if (isWildcard) {
+      for (const decl of exportList) {
+        decl[1] = file
+        decl[2] = decl[0]
+      }
+    }
+
+    // This recursively handles the wildcard exports (e.g. `export * from './a'`)
+    if (wildcardExports.length) {
+      await Promise.all(
+        wildcardExports.map(async (req) => {
+          const targetPath = await resolve(
+            path.dirname(file),
+            req.replace('__barrel_optimize__?names=__PLACEHOLDER__!=!', '')
+          )
+
+          const targetMatches = await getMatches(targetPath, true)
+          if (targetMatches) {
+            // Merge the export list
+            exportList = exportList.concat(targetMatches.exportList)
+          }
+        })
+      )
+    }
+
+    return {
+      prefix,
+      exportList,
+      wildcardExports,
+    }
+  }
+
+  const res = await getMatches(resourcePath, false)
+  barrelTransformMappingCache.set(resourcePath, res)
+
+  return res
+}
+
 const NextBarrelLoader = async function (
   this: webpack.LoaderContext<{
     names: string[]
-    wildcard: boolean
+    swcCacheDir: string
   }>
 ) {
   this.async()
-  const { names, wildcard } = this.getOptions()
-
-  const source = await new Promise<string>((resolve, reject) => {
-    this.loadModule(
-      `__barrel_transform__${wildcard ? '?wildcard' : ''}!=!${
-        this.resourcePath
-      }`,
-      (err, src) => {
-        if (err) {
-          reject(err)
-        } else {
-          resolve(src)
-        }
-      }
-    )
+  this.cacheable(true)
+
+  const { names, swcCacheDir } = this.getOptions()
+
+  // For barrel optimizations, we always prefer the "module" field over the
+  // "main" field because ESM handling is more robust with better tree-shaking.
+  const resolve = this.getResolve({
+    mainFields: ['module', 'main'],
   })
 
-  const matches = source.match(
-    /^([^]*)export const __next_private_export_map__ = ('[^']+'|"[^"]+")/
+  const mapping = await getBarrelMapping(
+    this.resourcePath,
+    swcCacheDir,
+    resolve,
+    this.fs
   )
 
-  if (!matches) {
+  // `resolve` adds all sub-paths to the dependency graph. However, we already
+  // cached the mapping and we assume them to not change. So, we can safely
+  // clear the dependencies here to avoid unnecessary watchers which turned out
+  // to be very expensive.
+  this.clearDependencies()
```

Comment:
Did you intentionally skip the `contextDependency` we talked about yesterday?

Reply:
(Add a note for future viewer here - we aren't watching `node_modules` in our watchers in the first place, so there's no need to add them as dependencies here)